modules/mesh-task/variables.tf

```diff
@@ -207,3 +232,115 @@ variable "application_shutdown_delay_seconds" {
   EOT
   default     = null
 }
+
+variable "consul_ecs_config" {
+  type        = any
```

Comment:
I tried a [structural type](https://www.terraform.io/language/expressions/type-constraints#structural-types) to describe this, but that requires users pass all the nested fields here, even if they only want to specify a few. There is experimental support for optional fields and defaults in Terraform, but I didn't want to rely on that since it is not yet stable.

Instead, I use `type = any` and some validation to this variable to ensure users are passing the correct fields. It does not validate the types. I thought some immediate validation in Terraform is good feedback, and the full JSON Schema validation will occur at runtime.

Structural type example:

```hcl
type = object({
  service = object({
    enableTagOverride = bool
    weights = object({
      passing = number
      warning = number
    })
    namespace = string
  })
  proxy = object({
    meshGateway = object({
      mode = string
    })
    expose = object({
      checks = bool
      paths = list(object({
        listenerPort = number
        path = string
        localPathPort = number
        protocol = string
      }))
    })
  })
})
```

Reply:
We could also look into using a JSON Schema Terraform provider to validate this data. I'm not certain if any of them are well supported and widely used though.

Reply:
> We could also look into using a JSON Schema Terraform provider to validate this data.

Yeah, great idea. I was definitely interested in doing this. But there's only [one](https://registry.terraform.io/providers/xxxbobrxxx/jsonschema/latest) that I could find and it has not been updated. There's a slightly updated [fork](https://github.com/JeffAshton/terraform-provider-jsonschema) though, but it's still on Go 1.15 for example.

The other implication is users need to know to include any providers we use in their `required_providers` list, so I'm hesitant to add dependencies without a bit more consideration, and it didn't seem essential here.

I'd prefer we actually explore creating our own "helper" provider for things like validation and utilities. I don't think it'd be all that much work. And since it's all Golang, it would give us the ability to do some code sharing with our consul-ecs binary.

Reply:
I agree completely 👍  We can definitely do this as enhancement later